### ISSUE TITLE
Media type instead of schemes?

### ISSUE THREAD

**Issue by jackfirth** [package-url/purl-spec#6]
So as is, the PURL specification's use of URIs has some issues standing in the way of adoption by the wider Internet community:

- Multiple schemes that are part of a broader interconnected protocol with new schemes required whenever a new kind of package manager is supported. Scheme deployment is lengthy and expensive, often requiring custom code not just in consumers and producers but in proxies, gateways, network-level middleboxes, application-level routers, and any number of other agents.
- Improper use of fragments which schemes are explicitly not allowed to place syntax constraints on. A fragment identifies a "secondary resource" contained within a [retrieved representation](https://github.com/package-url/purl-spec/issues/5) of the primary resource. Because of it's dependence on the format of the representation, only format definitions ([IANA Media Types](https://www.iana.org/assignments/media-types/media-types.xhtml)) can specify how fragments are interpreted. A scheme therefore can only add meaning to fragments indirectly by restricting what media types are legal representations. But none of the PURL schemes define how to retrieve packages at all, let alone what media types can be retrieved.
- Overlap with existing definition of URI authorities. The README states: "A namespace segment may sometimes look like a host but its interpretation is specific to a type." But schemes are already allowed to define authorities with custom interpretation; there's no requirement that locator schemes define locations in terms of IP / DNS addresses. RFC 3986 even [explicitly defines](https://tools.ietf.org/html/rfc3986#section-3.2) authorities as governors of some namespace so there's clear overlap in purpose here.
- Most of all: **unclear utility to consumers**. What is a client supposed to *do* with a PURL? From use cases you've described to me, it seems the primary purpose of a PURL is to have a compact textual representation of various kinds of package data that generic package tools parse and use for comparison to verify they're talking about the "same" package. None of "dereferencing", "installing", "publishing", or "deleting" occurs with a PURL because those protocols are completely dependent on what kind of package is used, which generic PURL clients are trying to avoid.

To expand on the unclear utility point, here are the proposed use cases as I understand them:

- Cross-system metadata indexing to search and monitor packages by metadata like available versions, dependencies, contributors, etc. across multiple package managers (libraries.io)
- Vulnerability tracking to determine whether a package's set of possible transitive dependencies includes a known vulnerability and whether the version constraints of that dependency graph allow or prevent patching
- Other kinds of package-content-agnostic analysis tools, especially tools that look at the dependency graphs of package ecosystems

All of these use cases really just need a common format for *package metadata*, particularly one that allows testing for whether two representations refer to the "same" package. I don't think a whole new URI scheme is needed for that. Instead, make a media type that represents this metadata and use that type in existing schemes like HTTP. A good existing type to model yours after is the [RFC 7807 `application/problem+json` type](https://tools.ietf.org/html/rfc7807) which looks like this:

```json
Content-Type: application/problem+json
{
    "type": "https://example.com/probs/out-of-credit",
    "title": "You do not have enough credit.",
    "detail": "Your current balance is 30, but that costs 50.",
    "instance": "/account/12345/msgs/abc",
    "balance": 30,
    "accounts": ["/account/12345",
                 "/account/67890"]
}
```

You could define something similar for package metadata. You could also define a *second* format that's purely textual (e.g. `text/pkgmeta` to complement `application/pkgmeta+json`) and acts as a short human-friendly string. Having both standard structured representations and text representations means that only human-facing systems need to care about parsing, making it far easier for non-human-facing systems to implement the format and avoid encoding bugs.

Media type registrations are much easier to obtain than schemes and less impactful on the wider Internet. A vendor tree registration like `application/vnd.librariesio.pkgmeta+json` takes little more than an email; see [RFC 6838](https://tools.ietf.org/html/rfc6838) for details.

**Comment by pombredanne** [comment]
> But none of the PURL schemes define how to retrieve packages at all, let alone what media types can be retrieved.

Actually I think retrieval is fully defined (even though this is NOT the primary usage). See https://github.com/package-url/purl-spec/issues/5#issuecomment-347132121

**Comment by pombredanne** [comment]
@jackfirth we addressed the fully defined aspects of purls in #9
Unless you can come with a complete rework using media type that has all the same URL properties of what we have today, I am closing this now. Please reopen if you feel strongly about this and thank you ++ for all the feedback you did provide!
